atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -417,7 +426,12 @@ public InetSocketAddress apply(String endpoint) {
             }
 
             try {
-                return hostPool.runWithPooledResource(f);
+                V result = hostPool.runWithPooledResource(f);
+                InetSocketAddress host = hostPool.getHost();
+                if (blacklistedHosts.containsKey(host)) {
```

Comment:
the change doesn't have the semantics your comment says it does (that this only affects the all-hosts-blacklisted case) because of concurrency.

old way:
- we open up a ton of connections to ${faulty server}
- at least a few of these have excessive problems even after retries and blacklist ${faulty server}
- ${faulty server} stays blacklisted for a while
- ${faulty server} eventually comes back
general gist: we slowdown and have backpressure good routing against slow servers

new way:
- we open up a ton of connections to ${faulty server}
- at least a few of these have excessive problems even after retries and blacklist ${faulty server}
- a really simple query against ${faulty server} and subsequent very simple health check barely works and unblocks it immediately
- more connections go to the not-blacklisted ${faulty server}
- at least a few of these have excessive problems even after retries and blacklist ${faulty server}
- a really simple query against ${faulty server} and subsequent very simple health check barely works and unblocks it immediately
- more connections go to the not-blacklisted ${faulty server}...